virtualmachine/vsphere/vm.go

```diff
@@ -837,56 +836,163 @@ func DeleteTemplate(vm *VM) error {
-			case "Network":
-				err = vm.collector.RetrieveOne(vm.ctx, networkMor, []string{"name"}, &network)
-				if err != nil {
-					return nil, err
-				}
-				networkList[dc.Name()] = append(networkList[dc.Name()], network.Name)
+// GetNetworkInHost : Returns the networks in a host in a cluster
+func GetNetworkInHost(vm *VM) ([]map[string]string, error) {
+	var (
+		networkList []map[string]string
+		networkMap  map[string]string
+		network     mo.Network
+		portGroup   mo.DistributedVirtualPortgroup
+		opNetwork   mo.OpaqueNetwork
+		hsMo        mo.HostSystem
+	)
+	// set up session to vcenter server
+	if err := SetupSession(vm); err != nil {
+		return nil, err
+	}
+
+	dc, err := GetDatacenter(vm)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the cluster resource and its host, network and datastore
+	crMo, err := findClusterComputeResource(vm, dc, vm.Destination.DestinationName)
+	if err != nil {
+		return nil, err
+	}
+
+	// find the host in Destination.HostSystem
+	for _, host := range crMo.Host {
+		err = vm.collector.RetrieveOne(vm.ctx, host, []string{"name", "network"}, &hsMo)
+		if err != nil {
+			return nil, err
+		}
+		if hsMo.Name == vm.Destination.HostSystem {
+			break
+		}
+	}
+
+	// Add all the networks in host to network list
+	for _, networkMor := range hsMo.Network {
+		switch networkMor.Type {
+		case "Network":
+			err = vm.collector.RetrieveOne(vm.ctx, networkMor, []string{"name"}, &network)
+			if err != nil {
+				return nil, err
+			}
+			networkMap = map[string]string{"name": network.Name}
+		case "DistributedVirtualPortgroup":
+			err = vm.collector.RetrieveOne(vm.ctx, networkMor, []string{"name"}, &portGroup)
+			if err != nil {
+				return nil, err
+			}
+			networkMap = map[string]string{"name": portGroup.Name}
+		case "OpaqueNetwork":
+			err = vm.collector.RetrieveOne(vm.ctx, networkMor, []string{"name"}, &opNetwork)
+			if err != nil {
+				return nil, err
 			}
+			networkMap = map[string]string{"name": opNetwork.Name}
 		}
+		networkList = append(networkList, networkMap)
 	}
 	return networkList, nil
 }
 
+// GetDcNetworkList : returns a list of network in given datacenter
+// available-filters (map-keys): "hosts", "clusters".
+func GetDcNetworkList(vm *VM, filter map[string][]string) ([]map[string]string, error) {
+	networkMap := map[string]bool{}
+	networks := []map[string]string{}
+	// set up session to vcenter server
+	if err := SetupSession(vm); err != nil {
+		return nil, err
+	}
+	if _, ok := filter["clusters"]; !ok {
+		return nil, errors.New("Key 'clusters' is missing")
+	}
+
+	if _, ok := filter["hosts"]; !ok {
+		return nil, errors.New("Key 'hosts' is missing")
+	}
+
+	vm.Destination.DestinationType = "cluster"
+	for _, cluster := range filter["clusters"] {
+		vm.Destination.DestinationName = cluster
+		for _, host := range filter["hosts"] {
+			vm.Destination.HostSystem = host
+			networksInHost, err := GetNetworkInHost(vm)
+			switch err.(type) {
+			case ErrorObjectNotFound:
```

Comment:
This is to check if a destination [host in a cluster] exist or not. 'GetNetworkInHost' searches the destination in vcenter server and returned 'ErrorObjectNotFound', if destination does not exist.